File: apps/api/schemas/contradiction.py
from pydantic import BaseModel, Field, ConfigDict

class ConflictingStatementSchema(BaseModel):
    text: str
    page: int
    documentId: str

class CitationSchema(BaseModel):
    documentId: str
    documentName: str
    pageNumber: int
    snippet: str
    score: float | None = None

class ContradictionInsightSchema(BaseModel):
    id: str
    type: str  # timeline, statement, numerical, logical, requirement, entity
    severity: str  # low, medium, high, critical
    confidence: float
    summary: str
    explanation: str
    conflictingStatements: list[ConflictingStatementSchema] = Field(default_factory=list)
    citations: list[CitationSchema] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

class ContradictionTelemetrySchema(BaseModel):
    retrievalCount: int
    contradictionCount: int
    reasoningLatency: float
    orchestrationLatency: float
    providerLatency: float

File: packages/types/src/contradiction.ts
import { Citation } from "./chat.js";

export interface ConflictingStatement {
  text: string;
  page: number;
  documentId: string;
}

export interface ContradictionInsight {
  id: string;
  type: "timeline" | "statement" | "numerical" | "logical" | "requirement" | "entity";
  severity: "low" | "medium" | "high" | "critical";
  confidence: number;
  summary: string;
  explanation: string;
  conflictingStatements: ConflictingStatement[];
  citations: Citation[];
}
